cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/security/jwt/JwtLoginService.java

```diff
@@ -0,0 +1,197 @@
+ * validation that is expected to be stored in a PEM formatted file.</p>
+ * <p>This class implements {@link AbstractLifeCycle} which means it is a managed bean, its lifecycle will be managed
+ * by Jetty. It's {@link AuthorizationService} can also be an {@link AbstractLifeCycle} in which case it delegates to
+ * this class, so opening and closing connections should be done by implementing the {@link AbstractLifeCycle} interface's
+ * {@link #doStart()} and {@link #doStop()} methods respectively. For a simple example see
+ * {@link UserStoreAuthorizationService}.</p>
+ */
+public class JwtLoginService extends AbstractLifeCycle implements LoginService {
+
+  private final AuthorizationService _authorizationService;
+  private IdentityService _identityService = new DefaultIdentityService();
+  private final RSAPublicKey _publicKey;
+  private final List<String> _audiences;
+
+  public JwtLoginService(AuthorizationService authorizationService, String publicKeyLocation, List<String> audiences)
+      throws IOException, CertificateException {
+    this(authorizationService, readPublicKey(publicKeyLocation), audiences);
+  }
+
+  public JwtLoginService(AuthorizationService authorizationService, RSAPublicKey publicKey, List<String> audiences) {
+    _authorizationService = authorizationService;
+    _publicKey = publicKey;
+    _audiences = audiences;
+  }
+
+  @Override
+  protected void doStart() throws Exception {
+    super.doStart();
+    // The authorization service might want to start a connection or access a file
+    if (_authorizationService instanceof LifeCycle) {
+      ((LifeCycle) _authorizationService).start();
+    }
+  }
+
+  @Override
+  protected void doStop() throws Exception {
+    if (_authorizationService instanceof LifeCycle) {
+      ((LifeCycle) _authorizationService).stop();
+    }
+    super.doStop();
+  }
+
+  @Override
+  public String getName() {
+    return null;
+  }
+
+  @Override
+  public UserIdentity login(String username, Object credentials, ServletRequest request) {
+    if (!(credentials instanceof SignedJWT)) {
+      return null;
+    }
+    if (!(request instanceof HttpServletRequest)) {
+      return null;
+    }
+
+    SignedJWT jwtToken = (SignedJWT) credentials;
+    boolean valid;
+    try {
+      valid = validateToken(jwtToken, username);
+    } catch (ParseException e) {
+      JWT_LOGGER.warn(String.format("%s: Couldn't parse a JWT token", username), e);
+      return null;
+    }
+    if (valid) {
+      String serializedToken = (String) request.getAttribute(JwtAuthenticator.JWT_TOKEN_REQUEST_ATTRIBUTE);
+      UserIdentity rolesDelegate = _authorizationService.getUserIdentity((HttpServletRequest) request, username);
+      return getUserIdentity(jwtToken, serializedToken, username, rolesDelegate);
+    } else {
+      return null;
+    }
+  }
+
+  @Override
+  public boolean validate(UserIdentity user) {
```

Comment:
Hmm, good question. On a second thought probably it would make sense to check the expiry (as audiences and signature doesn't change for an issued user). I'll update this.